File: simulator/ts/geo/landingzone.ts
import { GeoPointV, LatLng, LatLngAlt, Point, Point3V, Turn } from "../dtypes"
import * as geo from "./geo"
import { toDegrees, toRadians } from "./trig"

/**
 * Defines landing zone parameters.
 * Also defines the coordinate system, centered on the landing zone, north aligned.
 */
export class LandingZone {
  public readonly destination: GeoPointV
  /** Landing direction in radians */
  public readonly landingDirection: number

  // Destination, as origin of coordinate system
  public readonly dest: Point3V

  constructor(destination: LatLngAlt, landingDirection: number) {
    this.destination = {
      ...destination,
      vN: 0,
      vE: 0,
      climb: 0,
      millis: 0
    }
    this.landingDirection = landingDirection
    this.dest = {
      x: 0,
      y: 0,
      alt: 0,
      vx: Math.sin(landingDirection),
      vy: Math.cos(landingDirection),
      climb: 0
    }
  }

  /**
   * Convert lat, lng to x, y meters centered at destination
   */
  public toPoint(ll: LatLng): Point {
    const bearing = toRadians(geo.bearing(this.destination.lat, this.destination.lng, ll.lat, ll.lng))
    const distance = geo.distance(this.destination.lat, this.destination.lng, ll.lat, ll.lng)
    return {
      x: distance * Math.sin(bearing),
      y: distance * Math.cos(bearing)
    }
  }

  /**
   * Convert lat, lng, alt msl to x, y meters, alt agl, centered at destination
   */
  public toPoint3V(ll: GeoPointV): Point3V {
    const bearing = toRadians(geo.bearing(this.destination.lat, this.destination.lng, ll.lat, ll.lng))
    const distance = geo.distance(this.destination.lat, this.destination.lng, ll.lat, ll.lng)
    return {
      x: distance * Math.sin(bearing),
      y: distance * Math.cos(bearing),
      alt: ll.alt - this.destination.alt,
      vx: ll.vE,
      vy: ll.vN,
      climb: ll.climb
    }
  }

  /**
   * Convert x, y coordinates to lat, lng
   */
  public toLatLng(point: Point): LatLng {
    const bear = Math.atan2(point.x, point.y)
    const dist = Math.sqrt(point.x * point.x + point.y * point.y)
    return geo.moveBearing(this.destination.lat, this.destination.lng, bear, dist)
  }

  /**
   * Convert x, y, alt agl coordinates to lat, lng, alt msl
   */
  public toLatLngAlt(point: Point3V): LatLngAlt {
    const bear = Math.atan2(point.x, point.y)
    const dist = Math.sqrt(point.x * point.x + point.y * point.y)
    return {
      ...geo.moveBearing(this.destination.lat, this.destination.lng, bear, dist),
      alt: point.alt + this.destination.alt
    }
  }

}

export const kpow = new LandingZone({lat: 47.239, lng: -123.143, alt: 84}, toRadians(32))
export const chelan = new LandingZone({lat: 47.864, lng: -119.944, alt: 378}, 0) // chelan
export const milehi = new LandingZone({lat: 40.162, lng: -105.164, alt: 1535}, 0) // mile hi

export const defaultLz = kpow

File: android/app/src/main/java/ws/baseline/paradrone/geo/LandingZone.java
package ws.baseline.paradrone.geo;

import ws.baseline.paradrone.GeoPoint;
import ws.baseline.paradrone.util.Convert;

import androidx.annotation.NonNull;
import java.util.Locale;

public class LandingZone {
    public static final LandingZone kpow = new LandingZone(47.239, -123.143, 84, Math.toRadians(32));

    @NonNull
    public final LatLngAlt destination;
    /** Landing direction in radians */
    public final double landingDirection;

    // Ground length of final approach
    public final double finalDistance = 150; // meters

    // Destination, as origin of coordinate system
    @NonNull
    public final Point3V dest;

    /**
     * Construct a LandingZone object.
     *
     * @param lat latitude in decimal degrees
     * @param lng longitude in decimal degrees
     * @param alt altitude msl in meters
     * @param landingDirection direction to be facing on landing in radians
     */
    public LandingZone(double lat, double lng, double alt, double landingDirection) {
        this.destination = new LatLngAlt(lat, lng, alt);
        this.landingDirection = landingDirection;
        this.dest = new Point3V(
                0,
                0,
                0,
                Math.sin(landingDirection),
                Math.cos(landingDirection),
                0
        );
    }

    /**
     * Convert lat, lng to x, y meters centered at current location
     */
    @NonNull
    public Point3V toPoint3V(@NonNull GeoPoint point) {
        final double bearing = Geo.bearing(this.destination.lat, this.destination.lng, point.lat, point.lng);
        final double distance = Geo.distance(this.destination.lat, this.destination.lng, point.lat, point.lng);
        return new Point3V(
                distance * Math.sin(bearing),
                distance * Math.cos(bearing),
                point.alt - destination.alt,
                point.vE,
                point.vN,
                point.climb
        );
    }

    /**
     * Convert x, y coordinates to lat, lng
     */
    @NonNull
    public LatLng toLatLng(@NonNull Point point) {
        final double bearing = Math.atan2(point.x, point.y);
        final double distance = Math.sqrt(point.x * point.x + point.y * point.y);
        return Geo.moveBearing(this.destination.lat, this.destination.lng, bearing, distance);
    }

    /**
     * Convert Point3V to GeoPointV
     */
    @NonNull
    public GeoPoint toGeoPoint(@NonNull Point3V point) {
        final LatLng ll = toLatLng(point);
        return new GeoPoint(ll.lat, ll.lng, point.alt, point.vy, point.vx, point.climb);
    }

    @NonNull
    @Override
    public String toString() {
        return String.format(Locale.getDefault(), "%.6f, %.6f, %s, %.0f°", destination.lat, destination.lng, Convert.distance(destination.alt), Math.toDegrees(landingDirection));
    }
}
